Add background hang monitor

gterzian opened the pull request:
<!-- Please describe your changes on the following line: -->

Early feedback welcome, still need to ~~actually integrate it in the constellation message loop, and~~ use it in Layout and Script...

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #16740 (github issue number if applicable).

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/21673)
<!-- Reviewable:end -->

gterzian committed:
introduce a background-hang-monitor:
Mac-Os implementation of a thread sampler,
Linux and Windows skeleton implementations.

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/constellation/background_hang_monitor.rs, components/constellation/constellation.rs, components/constellation/tests/hang_monitor_tests.rs, components/constellation/lib.rs
 * @wafflespeanut: python/servo/testing_commands.py
 * @cbrewster: components/constellation/background_hang_monitor.rs, components/constellation/constellation.rs, components/constellation/tests/hang_monitor_tests.rs, components/constellation/lib.rs
 * @paulrouget: components/constellation/background_hang_monitor.rs, components/constellation/constellation.rs, components/constellation/tests/hang_monitor_tests.rs, components/constellation/lib.rs
 * @kichjang: components/script_traits/script_msg.rs

bors-servo commented:
:umbrella: The latest upstream changes (presumably #21672) made this pull request unmergeable. Please resolve the merge conflicts.

gterzian commented on components/script/script_thread.rs:
```diff
@@ -1035,6 +1037,16 @@ impl ScriptThread {
+                .map(|id, _| MonitoredComponentType::Script(id.clone()))
+                .collect();
+            let msg = MonitoredComponentMsg::NotifyWait(waiting_components);
+            let _ = self.script_sender.send((id.clone(), ScriptMsg::ForwardToBackgroundHangMonitor(msg)));
```
Yes that's a good idea(I looked at using the `TopLevelBrowsingContextId`, and since we introduced auxiliaries  a script-thread could also manage multiple of those. 

I can add a new Id to the existing `EventLoop` concept.

gterzian replied:
An event-loop id might also not well reflect the reality, since while we indeed will have one script-thread per event-loop, we can have several layout threads. For now I have opted for this https://github.com/servo/servo/pull/21673/files#diff-8dc9268416cd4f037d50206e76c73cf9R525

gterzian commented on components/script/script_thread.rs:
```diff
@@ -1239,6 +1254,14 @@ impl ScriptThread {
 
+    fn notify_activity_to_hang_monitor(&self, category: &ScriptThreadEventCategory, pipeline_id: &PipelineId) {
+        let position = ScriptThreadEventCategory::into_enum_iter().position(|x| &x == category).unwrap();
+        let hang_annotation = ScriptHangAnnotation::into_enum_iter().nth(position).unwrap();
```
I agree, I will look into it...  could be useful to apply to https://github.com/servo/servo/blob/cd02ca6c191f61865eafb4c62af532b19db53030/components/script/script_thread.rs#L1286 as well

gterzian commented on components/constellation/constellation.rs:
```diff
@@ -1095,6 +1106,9 @@ where
         };
 
         match content {
+            FromScriptMsg::ForwardToBackgroundHangMonitor(msg) => {
```
I agree.

gterzian commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 4e928985f180420a42d4b3fd2ee55115258ed3b4 with merge cd3b0dd3ed88d7a9d5cd2b3917fd108aa35077cc...

gterzian commented:
@jdm I've restructured the whole thing to better match the script and layout thread structure, so we now have one monitoring thread per layout or script, and I've also added a very rudimentary form of backtrace via a signal send to the process by the monitor thread in case of a permanent hang, unfortunately Ithink it will be handled by all threads in the process, so you'll get a backtrace not just of the hanging one, I guess it's a start... :)

gterzian commented:
I'll try to target the signal to the hanging thread specifically using `libc`.

On another point, I did some testing and it seems to backtrace only goes so far as the signal handler, and does not show "what the thread was doing while receiving the signal". It might be because I am using `thread::sleep` to simulate a hang? I'm not sure and I expected the backtrace to include more than just the signal handler...

gterzian commented:
I'll try to target the hanging thread more specifically using `libc::p_thread` and company, however I'm still figuring out how to get a full stack trace beyond the code run in the signal handler. 

this provides a good description of how the linux part is done in Gecko: https://dxr.mozilla.org/mozilla-central/source/tools/profiler/core/platform-linux-android.cpp?q=SuspendAndSampleAndResumeThread&redirect_type=direct#148

gterzian commented:
@jdm I think this one could use another round of review. I've added some code pertaining to the `SIGPROF` handling and also actually trying to communicate the backtrace back to the monitoring thread(who then passes it on  to the constellation). 

It's a bit messy and I'm not sure if it's actually a good idea... :)

gterzian commented:
Ok I cleaned up the traceback stuff a bit, and the good news is I was able to get the whole thing to crash in the tests due to concurrent access to the `static`(I had made an error with the return value of `compare_and_swap`) and fixed it. So it would appear to be somewhat more robust now...

gterzian commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 0d84ce3e60f834c108bb8428b02dc4d01c8b2abc with merge 3e34206daa2d50d02a5168266d7b62aa34d0e65a...

bors-servo commented:
:broken_heart: Test failed - [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/8308)

gterzian commented:
Ok, 

1. using `libc` doesn't work on windows
2. Unexpected dynamic symbols in binary: signal
3. The build time is crazy on Mac.

I've noticed 3 locally as well, by the way, memory usage during build would skyrocket to the point of freezing my Mac...

jdm commented on components/msg/background_hang_monitor.rs:
```diff
@@ -0,0 +1,191 @@
+                let mut component = self
+                    .monitored_components
+                    .get_mut(&component_id)
+                    .expect("Receiced NotifyActivity for an unknown component");
```
nit: Received

jdm commented on components/msg/background_hang_monitor.rs:
```diff
@@ -0,0 +1,191 @@
+            }
+            let last_annotation = monitored.last_annotation.unwrap();
+            if monitored.last_activity.elapsed() > monitored.permanent_hang_timeout {
+                match monitored.sent_permanent_alert {
```
```rust
if monitored.sent_permanent_alert {
    continue;
}
...
```

jdm commented on components/msg/background_hang_monitor.rs:
```diff
@@ -0,0 +1,191 @@
+            }
+            if monitored.last_activity.elapsed() > monitored.transient_hang_timeout {
+                match monitored.sent_transient_alert {
+                    true => continue,
```
Same.

jdm reviewed (commented):
I like a lot of what I'm seeing here! I'm mostly concerned about the large number of extra threads.

gterzian commented:
@jdm Ok I have addressed your comments:

- One monitor thread per process(or just one if not in multiprocess mode). This somewhat follows the pattern of ServiceWorkerManager.
- I've tried using the atomic crate, however there were a couple of problems:
    1. `Atomic<T>` required `T` to be Copy, so it was quite hard to get the backtrace across.
    2. On my machine `is_lock_free` would be false, indicating that the atomic was implemented with locks, which in our case seems to defeat the purpose of using it in the first place, right?

So I simply used a combo of an `mut Option`, and an `AtomicBool` to synchronize access to it(and since there won't be multiple monitoring thread, there is no need for the transaction concept anymore).

gterzian commented on components/msg/background_hang_monitor.rs:
```diff
@@ -0,0 +1,183 @@
+            recv(self.port.select(), event) => {
+                match event {
+                    Some(msg) => Some(msg),
+                    // Our sender has been dropped, quit.
```
I'm wondering if the assumption that the sender will be dropped is still valid, although I think it is. In single process mode even the constellation will own a sender, yet that will be dropped when the constellation exit, right?

gterzian commented:
It's also noteworthy that in Gecko they seem to have a more complicated "messaging" structure between monitor and monitored thread, apparently in order to do most of the stack sampling work from within the monitor, and only resume the monitored when that is done. 

We just get the stack from inside the signal handler and pass it on to the monitor. 

I wonder if we should look into taking the stacktrace from within the monitor instead...

See https://dxr.mozilla.org/mozilla-central/source/tools/profiler/core/platform-linux-android.cpp?q=SuspendAndSampleAndResumeThread&redirect_type=direct#156

gterzian commented:
Perhaps we could use the `backtrace::trace` functionality, and pass an array of of `(ip, symbo_addess)`(both are `*mut c_void`) to the monitor via a static variable, then resolve the actual backtrace from inside the monitor thread, leaving the other thread suspended, and when the monitor is done finally resume the other thread? 

See https://docs.rs/backtrace/0.3.9/backtrace/struct.Frame.html

It would be somewhat similar to what Gecko is doing, while avoiding too much new low-level code?

alexcrichton commented:
FWIW both `libbacktrace` as compiled by `backtrace` as well as `backtrace` are not safe to use in signal handlers, they'll both segfault and deadlock in obscure ways. I wouldn't recommend using either of those libraries from a signal handler.

gterzian commented:
As a quick update, it appears that I have been sucked into some kind of time-absorption vortex, and have yet to emerge from it.

I hope to get back to a more stable routine within the next few weeks, and to be able to focus on this problem again.

gterzian commented:
@jdm Could I please ask you to take a look at https://github.com/servo/servo/pull/21673/commits/aaa5e84af0b5542653927158e8b064ad52aaa729#diff-32c6131016718303d2824e09234521b1R172

and compare it with https://dxr.mozilla.org/mozilla-central/source/mozglue/misc/StackWalk.cpp#888

I could use some pointers(no pun intended) there because while the Gecko implementation looks deceivingly simple, I think I'm missing something fundamental when it comes to manipulating those pointers so as to walk down the stack...

The good news is that the last frame seems to resolve fine(it prints as `___semwait_signal`, which I assume corresponds to the the call to `thread::sleep`), so I think it's really the stack walk that needs fixing and the rest seems ok.

gterzian commented:
@jdm Ok I think the last commit is good for an intermediary review. 

On mac, I'm able to get a pretty decent backtrace looking like: `["_thread_start", "__pthread_start", "__pthread_body", (...) hang_monitor_tests20test_hang_monitoring (...) unix6thread6Thread5sleep (...) "___semwait_signal"]`...

jdm commented on components/msg/background_hang_monitor.rs:
```diff
@@ -2,50 +2,244 @@
+    }
+}
+
+fn symbolize_backtrace(native_stack: &mut NativeStack, hang_profile: &mut Option<HangProfile>) {
```
Same comment about returning a value instead of accepting a mutable argument. Additionally, since we only return Some, we can drop the Option.

jdm commented on components/msg/background_hang_monitor.rs:
```diff
@@ -2,50 +2,244 @@
+        backtrace: Vec::new()
+    };
+    native_stack.instruction_ptrs.reverse();
+    for ip in native_stack.instruction_ptrs.iter() {
```
We can use `native_stack.instruction_ptrs.iter().rev()` instead.

jdm commented on components/msg/background_hang_monitor.rs:
```diff
@@ -2,50 +2,244 @@
+}
+
+#[allow(unsafe_code)]
+unsafe fn frame_pointer_stack_walk(regs: &mut Registers, native_stack: &mut NativeStack) {
```
Does regs need to be mutable?

gterzian replied:
The reason I am passing all those `&mut` around as function argument, is because I wanted to avoid any potential allocation inside the "critical section". To be honest I have only limited understanding of the topic. 

So for example creating and returning a bool inside a function, would that not be doing a 'dynamic allocation'? And for those registers, I could just create them here and return the value, yet I also was "trying to prevent a dymamic allocation". 

I would off-course prefer not to pass around mutable references and instead just be returning values. Could you pls confirm that this would be ok to do in the "critical section", while the samplee thread is suspended?

jdm replied:
Dynamic allocation refers to anything that involves Vec, Box, Arc, Rc, etc. Anything that is stack allocated (such as a fixed-size array, or a bool, or a struct that doesn't contain types that require dynamic allocation) are totally fine to pass around and return.

gterzian replied:
Great! thanks for the explanation!

jdm commented on components/msg/background_hang_monitor.rs:
```diff
@@ -2,50 +2,244 @@
+    let kret = mach::thread_act::thread_get_state(thread_id,
+                                                  mach::thread_status::x86_THREAD_STATE64,
+                                                  mem::transmute(&state),
+                                                  mem::transmute(&state_count));
```
These are transmuting immutable pointers to mutable ones; the values would be declared with `mut` and we should pass &mut to the first transmute and have no need of the second transmute.

jdm commented on components/msg/constellation_msg.rs:
```diff
@@ -458,8 +469,9 @@ pub fn start_monitoring_component(
     let component_id = MonitoredComponentId(pipeline, component_type);
     let bhm_chan = BackgroundHangMonitorChan::new(sender, component_id.clone());
     let thread_id = unsafe {
+        #[cfg(any(target_os = "android", target_os = "linux"))]
```
A lot of this code would be nicer to modify outside of the msg crate. We might be able to define something like:
```rust
trait BackgroundHangMonitor {
    type ThreadId: Into<usize>;
    fn get_thread_id() -> Self::ThreadId;
    fn start_monitoring_component(...);
}
```
and move the majority of the implementation into a crate that only has components/servo depend on it (and include the BackgroundHangMonitor trait in the generics for ScriptThread and LayoutThread).

gterzian replied:
Good idea, let me see what I can do...

jdm reviewed (changes requested):
Using the backtrace library to resolve the symbols is a clever choice!

jdm commented on components/msg/background_hang_monitor.rs:
```diff
@@ -82,63 +80,66 @@ impl NativeStack {
+    check_kern_return(kret)?;
+    let thread_state = *(state as *mut mach::structs::x86_thread_state64_t);
+    Ok(Registers {
+        instruction_ptr: thread_state.__rip as Address,
```
Let's do this:
```rust
let mut state = mach::structs::x86_thread_state64_t::new();
let mut state_count = mach::structs::x86_thread_state64_t::count();
let kret = mach::thread_act::thread_get_state(thread_id,
                                                  mach::thread_status::x86_THREAD_STATE64,
                                                  (&mut state) as *mut _ as *mut _,
                                                  &mut state_count);
    check_kern_return(kret)?;
   Ok(Registers {
        instruction_ptr: state.__rip as Address,
        ...
    })
```

jdm commented on python/servo/build_commands.py:
```diff
@@ -188,10 +188,14 @@ class MachCommands(CommandBase):
                      default=None,
                      action='store_true',
                      help='Build the libsimpleservo library instead of the servo executable')
+    @CommandArgument('--with_frame_pointer',
```
`--with-frame-pointer`

bors-servo commented:
:umbrella: The latest upstream changes (presumably #21644) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
That looks like a great improvement!

bors-servo commented:
:umbrella: The latest upstream changes (presumably #22074) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
@gterzian Are you waiting on review at this point? Does the WIP in the title still make sense?

gterzian commented:
@jdm thanks for asking. I was actually pondering what to do next. 

Here is the current state of things:

1. Mac os 64 bit seems to work. This implementation uses native `mach` commands. 
2. For Linux/Android(Gecko combines those into implementation) using `libc`, the suspending/resuming, and sending the context across, works, however I was having problems with getting the instruction and frame pointers out of the context. This, and perhaps also the stack walking, might require a platform specific implementation. 
3. Windows is only sketched with empty functions, with TODOs pointing to potentially useful `winapi` commands. 
4. For both 2 and 3, the incompleteness doesn't prevent the overall system from functioning, the backtrace will simply be `None`, we will still get the hang alert only containing an event annotation.

I was thinking that the right strategy might be trying to merge it as it is, and do 2 and 3 incrementally, probably with the help of different people who have convenient access to these platforms? A blog post and drumming-up of help on reddit might be a good idea...

What do you think?

jdm commented:
That sounds like a sensible plan to me!

gterzian commented:
Ok, this one is ready for review then. The "needs rebase" label could be removed...

jdm commented on components/constellation/pipeline.rs:
```diff
@@ -295,7 +306,16 @@ impl Pipeline {
                     let _ = unprivileged_pipeline_content.spawn_multiprocess()?;
                 } else {
-                    unprivileged_pipeline_content.start_all::<Message, LTF, STF>(false);
+                    // Should not be None unless in multiprocess mode.
```
Let's use `expect()` for this instead, in that case.

jdm commented on components/msg/Cargo.toml:
```diff
@@ -13,9 +13,15 @@ test = false
 malloc_size_of_derive = { path = "../malloc_size_of_derive" }
 serde = "1.0.60"
+servo_channel = {path = "../channel"}
+sig = "0.1"
```
Are these all still necessary after moving the code to the new crate?

jdm commented on components/msg/constellation_msg.rs:
```diff
@@ -285,3 +287,184 @@ pub enum InputMethodType {
+                write!(fmt, " - <unknown>")?;
+            }
+
+            if let (Some(ref file), Some(line)) = (symbol.filename.clone(), symbol.lineno) {
```
Use `symbol.filename.as_ref()` instead of cloning.

jdm commented on components/script/script_thread.rs:
```diff
@@ -1126,6 +1148,9 @@ impl ScriptThread {
         let mut sequential = vec![];
 
+        // Notify the background-hang-monitor we are waiting for an event.
+        self.background_hang_monitor.notify_wait();
```
Let's add a call to `notify_activity_to_hang_monitor` inside of `ScriptThread::profile_event`, which will take care of a bunch of cases below this where we are actually not waiting.

jdm reviewed (changes requested):
This looks really good! There are some subtleties around the concurrency which we need to think about a bit more, but otherwise I think we're almost there!

Aaron1011 commented on components/background_hang_monitor/sampler.rs:
```diff
@@ -0,0 +1,337 @@
+
+#[allow(unsafe_code)]
+pub unsafe fn suspend_and_sample_thread(thread_id: MonitoredThreadId) -> Option<HangProfile> {
+    // Warning: The "critical section" begins here.
```
I think panicking is going to need special handling inside the critical section.
The default panic handler will allocate memory and try to print a backtrace, which we definitely don't want to do in the critical section.

I think the best solution would be to use `std::panic::set_hook` to register a custom panic handler which unconditionally aborts. The old panic handler can be retrieved with `std::panic::take_hook`, and restored once the critical section has ended.

I don't think that using `std::panic::catch_unwind` will work, since it allocates a `Box` to hold the error.

gterzian commented on components/constellation/constellation.rs:
```diff
@@ -586,6 +600,18 @@ where
+                    route_ipc_receiver_to_new_mpsc_receiver_preserving_errors(ipc_bhm_receiver);
+
+                let background_monitor_register = match opts::multiprocess() {
+                    true => None,
```
It is supported in multiprocess mode, in which case it is set to `None` here, and will instead be initiated at https://github.com/servo/servo/pull/21673/files/2dde0a03bffe09b11f9ddd7c95cf4771bef59e2c#diff-da41bfaea1380168143603c085efe1c3R646

jdm commented on components/background_hang_monitor/sampler.rs:
```diff
@@ -0,0 +1,337 @@
+    static ref SAMPLEE_CONTEXT_READY: AtomicBool = AtomicBool::new(false);
+    /// A flag used to signal when the samplee may resume executing,
+    /// and also used to know when it has resumed.
+    static ref SAMPLEE_MAY_RESUME: AtomicBool = AtomicBool::new(false);
```
Oh, right, there's only a single thread that is initiating the sampling at any given time. That does negate my argument about the racing, but I agree that a single state machine would be clearer than multiple boolean flags.

gterzian commented on components/background_hang_monitor/sampler.rs:
```diff
@@ -0,0 +1,337 @@
+
+/// The means of communication between monitored and monitor.
+#[cfg(any(target_os = "android", target_os = "linux"))]
+static mut SAMPLEE_CONTEXT: Option<(MonitoredThreadId, libc::mcontext_t)> = None;
```
I've tried using the "atomic" crate, and it turned out that, on my Mac at least, those would end up using locks internally(there is a `Atomic::<T>::is_lock_free()` function that returned true on my machine). 

So the idea was instead to use a set of `AtomicBool` flags to regulate access to this unlocked data. It looks like those flags will now be replaced by a statemachine encoded in a `AtomicInt`...

Aaron1011 replied:
The issue is that without atomic reads/writes, you have no guarantee that the sampler thread will see the write made by the samplee.

Also, the `atomic` crate appears to provide atomic access to an entire type, instead of a pointer to it. I think the [atom](https://crates.io/crates/atom) would work well - it's a thin wrapper that allows you to easily use a `Box`ed type with `AtomicPtr`.

gterzian replied:
Thanks, I will look into the "atom" crate. 

The goal is to implement similar to https://dxr.mozilla.org/mozilla-central/source/tools/profiler/core/platform-linux-android.cpp#148

which indeed seems to use a pointer to copy across the register state, see https://dxr.mozilla.org/mozilla-central/source/tools/profiler/core/platform-linux-android.cpp#215

Aaron1011 replied:
I think this still suffers from the same problem - since it's not accessed atomically, the sampler thread may not see writes made by the samplee thread.

gterzian replied:
@jdm what do you think? 

I think the state machine, and the fact that this back and forth is limited to two threads, indirectly guarantees it's own kind of "atomicity".

jdm replied:
I think Aaron1011 has a point. The thread could perform the non-atomic write in such a way that when the other thread tries to read the value, it sees a different value instead: https://preshing.com/20130618/atomic-vs-non-atomic-operations/

gterzian replied:
Thanks for the article. Could our use of `Ordering::SeqCst` prevent a data race problem? Having just [read more about it](https://doc.rust-lang.org/nomicon/atomics.html#sequentially-consistent), it appears to me that the read/write to the non-atomic data is "safe" thanks to the following:

1. Monitor loops and waits for `ThreadProfilerState::ReadyToObtainBacktrace` using `Ordering::SeqCst`.
2. Samplee writes the data. 
3. Samplee sets `ThreadProfilerState::ReadyToObtainBacktrace`, also using `Ordering::SeqCst`.
4. Monitor reads the data. 

From what I understand about "SeqCst access", is that it would prevent 2 and 4 above, which are indeed non-atomic in themselves, to be re-ordered in any way so that they would not occur in the order you would expect from ready the code. 

Please let me know what you think of this logic...

gterzian replied:
Here is another related discussion: https://stackoverflow.com/questions/35153905/least-restrictive-memory-ordering-for-spin-lock-with-two-atomics?rq=1

I think we're fine. If anything, a future optimization might want to switch to aquire/release ordering as opposed to doing everything in `Ordering::SeqCst`.

What do you think?

gterzian replied:
I've tried to make the workflow clearer with https://github.com/servo/servo/pull/21673/commits/8e65cf16e1abbce71cd2280d40e0e94fc604071c

Aaron1011 replied:
Unfortunately, I think that part of the documentation is incorrect - there's currently an open PR to change it: https://github.com/rust-lang/rust/pull/56023. `SeqCst` only prevent other `SeqCst` operations from being reordered - it doesn't affect non-atomic loads.

See also the [LLVM docs](https://llvm.org/docs/Atomics.html#sequentiallyconsistent) - `SeqCst` only provides guarantees about atomic operations.

gterzian replied:
Ok thanks for pointing this out. If the "sequential" part only applies to the actual `load` and `store`, and not to whatever code happens to be sandwitched in between them, then I agree we'll need to wrap the shared data into it's own variant of atomic(maybe an `AtomicPtr` or indeed using the [`atom`](https://docs.rs/atom/0.3.5/atom/) crate. 

My impression from reading about memory ordering was instead that if you used `SeqCst`, it creates a kind of barrier that also makes the surrounding code sequentially consistent. 

@stjepang If you have a moment, could you please take a look at the following:

1. We're operating inside a "critical section"(a thread is currently suspended), and cannot use locks or heap allocation. 
2. We have two threads needing to do a workflow inside this section, including writing and reading to a piece of shared data.
3. We have a piece of shared data: `static mut SAMPLEE_CONTEXT: Option<(MonitoredThreadId, libc::mcontext_t)> = None;`
4. We have a state-machine used to coordinate the overall worflow, as well as the access to the shared data: 
```
static ref SAMPLER_STATE: AtomicUsize = AtomicUsize::new(ThreadProfilerState::NotProfiling as usize);

enum ThreadProfilerState {
    NotProfiling = 0,
    ReadyToWriteContext,
    ReadyToObtainBacktrace,
    ObtainedBacktrace,
    ResumedThread,
}
```
5. The actual workflow is implemented by storing and reading the ` AtomicUsize`, with the reading thread looping until the ` AtomicUsize` is in a specific state. 
6. All the stores and loads are done using `SeqCst`.
7. The shared data is written to by one thread, then that same threads does a `store` on the state machine. The other thread is doing a `load`, and waiting for that state, and only reading the data after the state machine has switched into it. 
8. This commit hopefully hightlights the overall flow: 8e65cf1

Question: Could the (non-atomic) reading and writing to the (unlocked) shared state, when sandwhiched in between storing and loading an atomic using `SeqCst`, still result in unexpected behavior?

gterzian replied:
@Aaron1011 I find this (un-merged) update to the docs to be consistent with my earlier statement: 

> An operation on an atomic variable can do three things:
* Make the atomic itself safe to use from multiple threads.
* Additionally, synchronize other (even non-atomic) memory. A store with [`Release`] ordering
forms a synchronization (unidirectional) edge with any thread that does a load with
[`Acquire`] *on the written value*.
*Finally, a [`SeqCst`] operation participates in a globally consistent time line 

(see https://github.com/rust-lang/rust/pull/55233/files#diff-9aa1eb154ee636dcdcbcb58bc4c6f6c1R182)

gterzian replied:
Finally, just looking at the docs:

> SequentiallyConsistent (seq_cst in IR) provides Acquire semantics for loads and Release semantics for stores. Additionally, it guarantees that a total ordering exists between all SequentiallyConsistent operations.
 
and 

> Acquire provides a barrier of the sort necessary to acquire a lock to access other memory with normal loads and stores.

and 

> Release is similar to Acquire, but with a barrier of the sort necessary to release a lock.

finally compare the above with 

> Monotonic is the weakest level of atomicity that can be used in synchronization primitives, although it does not provide any general synchronization. It essentially guarantees that if you take all the operations affecting a specific address, a consistent ordering exists.

@Aaron1011 Could it be that what you have in mind is this "Monotonic" model? This indeed only synchronizes the specific address inside the actual atomic. 

For the other definitions, I read them as such at `SequentiallyConsistent`, `Acquire`, and `Release` can be used to effectively lock other memory addresses than the one "contained" inside the atomic...

stjepang replied:
@gterzian 

> My impression from reading about memory ordering was instead that if you used SeqCst, it creates a kind of barrier that also makes the surrounding code sequentially consistent.

Like @Aaron1011 said, this is not true. Atomic `SeqCst` operations are weaker than `SeqCst` [fences](https://doc.rust-lang.org/nightly/std/sync/atomic/fn.fence.html) (that's the kind of barrier you probably meant).

> If you have a moment, could you please take a look at the following:

The steps you wrote seem correct. You could even use `Acquire` and `Release` instead of `SeqCst` (but don't have to).

Just to clarify something:

* Mutable shared data **must** be atomic. No exceptions.
* Immutable shared data doesn't need to be atomic.
* Non-shared data doesn't need to be atomic (even if it is mutable and/or sent to another thread).

So if you want to have a mutable shared `Option<(MonitoredThreadId, libc::mcontext_t)>`, you must mutate an atomic variable. You should wrap the data in a `Mutex`, `AtomicPtr`, `Atom`, or something similar.

Note that in `Mutex<T>`, the internal mutex state is mutable and shared, while the `T` inside is not shared. In `RwLock<T>`, the `T` is always either shared and immutable (during read locks) or non-shared and mutable (during write locks).

It's also important to say that in order to send any kind of non-atomic data to another thread, you should synchronize it through a `Release` write and an `Acquire` read.

So when you create a `(MonitoredThreadId, libc::mcontext_t)` and want to make it visible to another thread, you must *publish* it through a write to the atomic variable with `Release` ordering (or stronger). The other thread then must read it with `Acquire` ordering (or stronger) in order to properly *fetch* the *published* data.

gterzian replied:
@stjepang thanks for the explanatio, and @Aaron1011 thanks for persistently pointing out this problem!

I'll remove this code, related to the linux implementation, for now, and revisit in another PR. I might ask for your opinion again when the time comes...

gterzian commented on components/background_hang_monitor/sampler.rs:
```diff
@@ -0,0 +1,337 @@
+        return None;
+    };
+    let native_stack = match get_registers(thread_id) {
+        Ok(regs) => Some(frame_pointer_stack_walk(regs)),
```
I only know that in Gecko, this is also done inside the critical section, see https://dxr.mozilla.org/mozilla-central/source/tools/profiler/core/platform.cpp#4006

gterzian commented:
Ok, I think I have addressed all comments, including @Aaron1011 one on adding a custom panic handler during the critical section. I've also added an "abort" in case resuming the thread fails.

@jdm ready for another round of review...

bors-servo commented:
:umbrella: The latest upstream changes (presumably #22142) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
Go ahead and squash the commits into a set that makes sense!
@bors-servo delegate+

bors-servo commented:
:v: @gterzian can now approve this pull request

gterzian commented:
Ok so in the light of https://github.com/servo/servo/issues/22203#issuecomment-440792016

I'm going to throw away the current linux code, and restructure mac to be in it's own `mac_os_sampler.rs` file, or something similar(maybe add a separate `frame.rs` file, since that part can be shared across platforms). 

Then in a separate PR, I'll re-add a linux implementation based on https://bitbucket.org/nikhilm/vignette/src/master/src/lib_linux.rs?fileviewer=file-view-default#lib_linux.rs-29.

So this will only affect `sampler.rs`, the rest of the code can stay the same...

jdm commented:
@bors-servo r+
Thanks!

bors-servo commented:
:pushpin: Commit 8393a62 has been approved by `jdm`

<!-- @bors-servo r=jdm 8393a62ae7a757f802ad9594a6cf0db64785e824 -->

bors-servo commented:
:hourglass: Testing commit 8393a62ae7a757f802ad9594a6cf0db64785e824 with merge 839679c8abff8bec03a6ba7f2570834f9b239cbc...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/6045)

jdm commented:
```
▶ TIMEOUT [expected OK] /_webgl/conformance/ogles/GL/asin/asin_001_to_006.html
  │ 
  │ VMware, Inc.
  │ softpipe
  │ 3.3 (Core Profile) Mesa 18.3.0-devel
  │ 2018-11-23 18:12:01.085 servo[78259:417598431] Metadata.framework [Error]: couldn't get the client port
  │ Stack trace
  │ stack backtrace:
  │    0:        0x10a3eff9e - __ZN9backtrace9backtrace5trace17h7e39d6bd7c678f2eE
  │    1:        0x10a3f0cac - __ZN72_$LT$backtrace..capture..Backtrace$u20$as$u20$core..default..Default$GT$7default17ha5434a657e6f38c1E
  │    2:        0x10a3f0d2d - __ZN9backtrace7capture9Backtrace3new17h027ac9deb1d86dcfE
  │    3:        0x107fe81e6 - __ZN5servo21install_crash_handler7handler17h3c15d5e0cedd67d2E
  │    4:     0x7fff87bb0529 - __sigtramp
  │    5:        0x10a3ef158 - __ZN112_$LT$background_hang_monitor..sampler_mac..MacOsSampler$u20$as$u20$background_hang_monitor..sampler..Sampler$GT$25suspend_and_sample_thread17h6e6f446a57eb6bcdE
  │    6:        0x10a3e8194 - __ZN23background_hang_monitor23background_hang_monitor27BackgroundHangMonitorWorker3run17h43b73fc5c553a031E
  │    7:        0x10a3e62f7 - __ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17hd41c2af7e08bd00aE
  │    8:        0x10a3e54b7 - __ZN3std9panicking3try7do_call17hb0f9f32e298ebf40E.llvm.9299181849444150992
  │    9:        0x10ae545ae - ___rust_maybe_catch_panic
  │   10:        0x10a3efadb - __ZN50_$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$8call_box17hae717c807f671d47E
  │   11:        0x10ae2bafb - __ZN3std3sys4unix6thread6Thread3new12thread_start17hba53abd905cebc29E
  │   12:     0x7fff98a1999c - __pthread_body
  │   13:     0x7fff98a19919 - __pthread_start
```

jdm commented:
The thing that actually concerned me was the sampler apparently crashed in the log I pasted.

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/6047)

gterzian commented:
@bors-servo try (with an even lower max frames)

bors-servo commented:
:hourglass: Trying commit 127c0434d131ef35916a7431d819372d69e83c7a with merge bdb5e3a97bf810c46a25da065fe8f26c3ea8680d...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/13231), [android-mac](http://build.servo.org/builders/android-mac/builds/542), [android-x86](http://build.servo.org/builders/android-x86/builds/1360), [arm32](http://build.servo.org/builders/arm32/builds/13394), [arm64](http://build.servo.org/builders/arm64/builds/13426), [linux-dev](http://build.servo.org/builders/linux-dev/builds/13533), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/10874), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/10926), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/13488), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/5976), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/5972), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/10244), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/10285), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/6048), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/5974), [magicleap](http://build.servo.org/builders/magicleap/builds/155), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/GpIco1pGQ1GqV-owrMIAuA), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/8876)
State: approved= try=True

gterzian commented:
Actually, I think this is less about lowering the "max frames", and more about ensure the stackwalk doesn't go above the highest address...

bors-servo commented:
:hourglass: Trying commit a0aaf0f8ec88ab665c33988cc36d5bb69a2dc005 with merge e8b82c872352bbbb5e69d7b7819b9f9b95b3c32e...

gterzian commented:
@jdm I prevented the crash by adding a check here: https://github.com/servo/servo/pull/21673/commits/63f893e20ecdd8be0b8d9a8113ec354f8eee5510#diff-71911b06e513640e808c4586db0ee8bcR108

![screen shot 2018-11-25 at 8 55 50 pm](https://user-images.githubusercontent.com/2792687/48979344-8b9feb00-f0f4-11e8-8e6d-c8a49bf2bfa1.png)


Tests all passed, I've fixed the tidy error since the test run

gterzian commented:
It seems that the backtrace is now too short with only one line, I'll have to look at the frame pointer stack walk again...

gterzian commented:
@bors-servo try  => looks like the earlier problem with the shortened backtrace is fixed

bors-servo commented:
:hourglass: Trying commit 897302e07294f0b56601d551ee798e9c4d78d3b5 with merge 1403163880af833cf6941a0e5922dc9c9cc06338...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/10878)

gterzian commented:
I think we might be sending too many "transient" hang alerts after a 100ms hang, which made some tests timeout? Changed it to 500ms...

gterzian commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 4227dce5f3d2b43030674176a26127335c98acd7 with merge 34fb8349f05ba90cc0da210bc6b678b1468c1b5a...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/10879)

gterzian commented:
Ok, I think there might be a problem with the suite, since other test runs unrelated to this PR also fail with a blank `filtered-wpt-errorsummary.log`...

gterzian commented:
@jdm what do you think would be good values for "transient" and "permanent" hangs?

 I've just set transient to 1s, and permanent to 5s. 

Previously tried setting transient to 100ms and 500ms, and that resulted in alerts on basically all page navigations. 

1s for transient doesn't seem to give alerts when browsing around github.com...

CYBAI commented:
@gterzian the failure in [linux-rel-css](https://build.servo.org/builders/linux-rel-css/builds/10879/steps/shell__4/logs/stdio) is related to servo/intermittent-failure-tracker#5

CYBAI commented:
@bors-servo try

- the issue from servo/intermittent-failure-tracker#5 has been fixed! Let's try again!

bors-servo commented:
:hourglass: Trying commit 4eb785cdc0446539bf5e7eb66bf7ad46ba5705dd with merge c6f4dcce47433b2677375c69b269c30028bbb82b...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/10886)

CYBAI commented:
Looks not related?

```
{"status": "CRASH", "group": "default", "message": null, "stack": null, "subtest": null, "test": "/_mozilla/mozilla/referrer-policy/no-referrer/meta-csp/cross-origin/http-https/iframe-tag/generic.no-redirect.http.html", "line": 218974, "action": "test_result", "expected": "OK"}
```

gterzian commented:
Yes those crashes look unrelated, also, they've occurred on Linux were we're not doing any sampling at the time. Glad to see the mac tests didn't crash this time...

gterzian commented:
@jdm OK I think it's good to do, the only significant change I made is https://github.com/servo/servo/pull/21673/files#diff-71911b06e513640e808c4586db0ee8bcR106 (as well as setting the 'transient' timeouts to 1000ms to prevent a flood of transient alerts).

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 4eb785c has been approved by `jdm`

<!-- @bors-servo r=jdm 4eb785cdc0446539bf5e7eb66bf7ad46ba5705dd -->

bors-servo commented:
:zap: Previous build results for [android](http://build.servo.org/builders/android/builds/13241), [android-mac](http://build.servo.org/builders/android-mac/builds/552), [android-x86](http://build.servo.org/builders/android-x86/builds/1370), [arm32](http://build.servo.org/builders/arm32/builds/13404), [arm64](http://build.servo.org/builders/arm64/builds/13436), [linux-dev](http://build.servo.org/builders/linux-dev/builds/13543), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/10939), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/13498), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/5987), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/5983), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/10257), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/10296), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/6059), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/5985), [magicleap](http://build.servo.org/builders/magicleap/builds/165), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/JdCiGfi-R-uSvuvF-bJIsw), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/8886) are reusable. Rebuilding only [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/10886)...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/13241), [android-mac](http://build.servo.org/builders/android-mac/builds/552), [android-x86](http://build.servo.org/builders/android-x86/builds/1370), [arm32](http://build.servo.org/builders/arm32/builds/13404), [arm64](http://build.servo.org/builders/arm64/builds/13436), [linux-dev](http://build.servo.org/builders/linux-dev/builds/13543), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/10891), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/10939), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/13498), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/5987), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/5983), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/10257), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/10296), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/6059), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/5985), [magicleap](http://build.servo.org/builders/magicleap/builds/165), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/JdCiGfi-R-uSvuvF-bJIsw), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/8886)
Approved by: jdm
Pushing c6f4dcce47433b2677375c69b269c30028bbb82b to master...
